tests/xtimer_now_irq/main.c

```diff
@@ -0,0 +1,47 @@
+/*
+ * Copyright (C) 2020 Freie Universität Berlin,
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     tests
+ * @{
+ *
+ * @file
+ * @brief       testing xtimer_now_usec function with irq disabled
+ *
+ *
+ * @author      Julian Holzwarth <julian.holzwarth@fu-berlin.de>
+ *
+ */
+
+#include <stdio.h>
+#include "xtimer.h"
+#include "irq.h"
+
+#define TEST_COUNT 4
+
+int main(void)
+{
+    puts("xtimer_now_irq test application.\n");
+
+    uint8_t count = TEST_COUNT;
+    while (count) {
+        unsigned int state = irq_disable();
+        uint32_t t1 = xtimer_now_usec();
+        xtimer_spin(xtimer_ticks((uint32_t)(~XTIMER_MASK) / 2));
```

Comment:
On 32bit 1mhz, this will spin `(0xffffffff / 2)` ticks. Doesn't that make the second iteration always fail?

1. xtimer_now() is somewhere a little higher than zero, so is t1
2. spin half period
3. t2 is now somewhere a little higher than a half period
4. next loop iteration, t1 is now close to t2 (a little higher than a half period)
5. spin half period
6. t2 overflows ("a little higher than a half period" + "half period" >> full period)
7. t2 < t1, error triggers

What am I missing?

Reply:
Running it as is for 32 bit timer is madness anyway and we need a better way for testing these corner cases with a bit of mocks magic.
Catching the "missed overflow" of the 32 bit timer with the 32 bit xtimer interface obviously doesn't make sense, the 64 bit variant is needed for that ;)

Reply:
Alright, how about straight passing if `XTIMER_WIDTH == 32`?

Reply:
Ahh just passing it would feel like lying. Why not just replace the `uint32_t t1 = xtimer_now_usec();` etc. with the 64 bit variants? Then it works for 32 bit too. 
Hmm I guess the problem is then we have no way to CI-run it for 16 bit timers *and* keep it working for manual 32 bit testing?.
How about an additional flag for manual running of 32 bit + some terminal output warning about 32 bit not being checked without that flag?

Reply:
> Ahh just passing it would feel like lying. Why not just replace the `uint32_t t1 = xtimer_now_usec();` etc. with the 64 bit variants? Then it works for 32 bit too.

yyyyyes, but isn't that testing something different then?

> How about an additional flag for manual running of 32 bit + some terminal output warning about 32 bit not being checked without that flag?

As long as by default on 32bit on CI, this just prints "not testing anything" "OK", so CI just passes. Could also blacklist.

As is it shouldn't run at all, on 32bit, just so noone actually waits 2**32usec for something to happen. :wink:

Reply:
> yyyyyes, but isn't that testing something different then?

Didn't test yet, but no not really

>As long as by default on 32bit on CI, this just prints "not testing anything" "OK", so CI just passes

I'd prefer some "not testing anything" "OK" over blacklisting.

> just so noone actually waits 2**32usec

I definitely had cases when testing timer stuff that made me desperate enough too run such long running tests^^